docker/manager.py

```diff
@@ -0,0 +1,821 @@
+        return value
+
+
+class TemplatesMixin(object):
+    """Mixin for generating Dockerfile from templates."""
+
+    _build_ctx = {}
+    # _build_path will consist of
+    #   key: package_tag
+    #   value: path to dockerfile
+    _build_path = {}
+
+    _template_env = Environment(
+        extensions=['jinja2.ext.do'], trim_blocks=True, lstrip_blocks=True
+    )
+
+    def __init__(self, release_spec, cuda_version):
+        # set private class attributes to be manifest.yml keys
+        for keys, values in release_spec.items():
+            super().__setattr__(f"_{keys}", values)
+
+        # setup default keys for build context
+        _update_build_ctx(self._build_ctx, self._release_spec, 'cuda')
+
+        # setup cuda_version and cuda_components
+        if self._check_cuda_version(cuda_version):
+            self._cuda_version = cuda_version
+        else:
+            raise RuntimeError(
+                f"CUDA {cuda_version} is either not supported or not defined in {self._cuda_dependencies.keys()}"
+            )
+        self._cuda_components = get_data(self._cuda_dependencies, self._cuda_version)
+
+    def __call__(self, dry_run=False, *args, **kwargs):
+        if not dry_run:
+            if FLAGS.python_version is not None:
+                logging.info(
+                    "--python_version defined, ignoring SUPPORTED_PYTHON_VERSION"
+                )
+                supported_python = [FLAGS.python_version]
+            else:
+                supported_python = SUPPORTED_PYTHON_VERSION
+
+            # we will remove generated dockerfile dir everytime generating new BentoML release.
+            logging.info(f"Removing dockerfile dir: {FLAGS.dockerfile_dir}")
+            shutil.rmtree(FLAGS.dockerfile_dir, ignore_errors=True)
+            try:
+                os.makedirs(FLAGS.dockerfile_dir)
+            except OSError as e:
+                if e.errno != errno.EEXIST:
+                    raise
+
+            self.generate_dockerfiles(
+                target_dir=FLAGS.dockerfile_dir, python_version=supported_python
+            )
+
+    @cached_property
+    def build_path(self):
+        return self._build_path
+
+    def _cudnn_versions(self):
+        for k, v in self._cuda_components.items():
+            if k.startswith('cudnn') and v:
+                return k, v
+
+    def _check_cuda_version(self, cuda_version):
+        # get defined cuda version for cuda
+        return True if cuda_version in self._cuda_dependencies.keys() else False
+
+    def _generate_distro_release_mapping(self, package):
+        """Return a dict containing keys as distro and values as release type (devel, cudnn, runtime)"""
+        releases = defaultdict(list)
+        for k, v in self._packages[package].items():
+            _v = list(flatten(v))
+            for distro in _v:
+                releases[distro].append(k)
+        return releases
+
+    def prepare_template_context(
```

Comment:
use ``--dump_metadata`` to see each image specs that will be parsed into our jinja templates.